sylt-common/src/value.rs

```diff
@@ -172,6 +160,115 @@ impl Value {
             Value::ExternFunction(slot) => slot + 2,
         }
     }
+
+    fn safe_fmt(
+        &self,
+        fmt: &mut std::fmt::Formatter<'_>,
+        seen: &mut HashSet<usize>
+    ) -> std::fmt::Result {
+        match self {
+            Value::Field(s) => write!(fmt, "<field .{}>", s),
+            Value::Ty(ty) => write!(fmt, "<type \"{:?}\">", ty),
+            Value::Blob(b) => write!(fmt, "<blob \"{}\">", b),
+            Value::Instance(_, v) => {
+                write!(fmt, "{} {{",
+                    if let Some(Value::String(name)) = v.borrow().get("_name") {
+                        name.as_str()
+                    } else {
+                        "?"
+                    }
+                )?;
+                if !seen.insert(self.unique_id()) {
+                    return write!(fmt, "...}}");
+                }
+                let mut first = true;
+                for e in v.borrow().iter() {
+                    if e.0.starts_with("_") {
+                        continue;
+                    }
+                    if !first {
+                        write!(fmt, ", ")?;
+                    }
+                    write!(fmt, "{}", e.0)?;
+                    write!(fmt, ": ")?;
+                    e.1.safe_fmt(fmt, seen)?;
+                    first = false;
+                }
```

Comment:
I _think_ this can be an iterator that filtermaps to `String` and uses https://doc.rust-lang.org/std/primitive.slice.html#method.join.

Reply:
It would use `format!` which explicitly allocates but IIRC `write!` already allocates.

Reply:
`join` would remove the trailing `, ` as well.

Reply:
I think we tried this - but I think the borrow checker gets sad about `fmt`